src/runner/browser-env/vite/browser-modules/driver.ts

```diff
@@ -0,0 +1,173 @@
+    const commands = new Set<string>();
+
+    for (const protocol of protocols) {
+        for (const endpointData of Object.values(protocol)) {
+            for (const { command } of Object.values<{ command: string }>(endpointData)) {
+                commands.add(command);
+            }
+        }
+    }
+
+    return Array.from(commands);
+}
+
+function getMockedProtocolCommands(): PropertiesObject {
+    return [...getAllProtocolCommands(), ...COMMANDS_HANDLED_BY_SERVER, ...window.__testplane__.customCommands].reduce(
+        (prev, commandName) => {
+            prev[commandName] = { value: mockCommand(commandName) };
+            return prev;
+        },
+        {} as PropertiesObject,
+    );
+}
+
+function mockCommand(commandName: string): ProtocolCommandFn {
+    return async (...args: unknown[]): Promise<unknown> => {
+        const { socket } = window.__testplane__;
+        const timeout = getCommandTimeout(commandName);
+
+        try {
+            // TODO: remove type casting after https://github.com/socketio/socket.io/issues/4925
+            const [error, result] = (await socket
+                .timeout(timeout)
+                .emitWithAck(BrowserEventNames.runBrowserCommand, { name: commandName, args })) as [
+                err: null | Error,
+                result?: unknown,
+            ];
+
+            if (error) {
+                throw error;
+            }
+
+            return result;
+        } catch (err) {
+            let error = err as Error;
+
+            if (error.message === SOCKET_TIMED_OUT_ERROR) {
+                error = new Error(
+                    `Didn't receive response from worker when executing browser command "${commandName}" with arguments "[${normalizeCommandArgs(
+                        commandName,
+                        args,
+                    )}]" in ${timeout}ms`,
+                );
+            }
+
+            throw error;
+        }
+    };
+}
+
+function getEnvironmentFlags(): Record<string, PropertyDescriptor> {
+    const { capabilities, requestedCapabilities, config } = window.__testplane__;
+
+    const environment =
+        config.automationProtocol === "devtools"
+            ? devtoolsEnvironmentDetector(capabilities as WebdriverIO.Capabilities)
+            : sessionEnvironmentDetector({
+                  capabilities,
+                  requestedCapabilities: requestedCapabilities as WebdriverIO.Capabilities,
+              });
+
+    return getEnvironmentVars(environment);
+}
+
+function getCommandTimeout(commandName: string): number {
+    const { config } = window.__testplane__;
+
+    let timeout = config.httpTimeout;
+
+    if (commandName === "debug" || commandName === "switchToRepl") {
+        timeout = SOCKET_MAX_TIMEOUT;
+    }
+
+    if (commandName === "navigateTo" && config.urlHttpTimeout) {
+        timeout = config.urlHttpTimeout;
+    }
+
+    return timeout;
+}
+
+function normalizeCommandArgs(name: string, args: unknown[] = []): unknown[] {
```

Comment:
Get implementation from server side, but rewrite a little bit in order to not use lodash.
I can't just import it from server side, because client code builds separately and knows nothing about nodejs code